Full user authentication

killlercram opened the pull request:
### Summary

This PR implements a complete user authentication system, including account creation, session management, and password recovery. It also integrates Google OAuth and token-based verification across both backend and frontend.

### Details

#### Backend Functionality

- Created `User` and `UserVerification` Sequelize models
- Implemented core auth routes:
  - `POST /api/users/signup` — Register user and send email verification
  - `POST /api/users/login` — Validate credentials and create JWT session
  - `POST /api/users/forgot-password` — Send reset link with secure token
  - `POST /api/users/reset-password` — Reset password using token
  - `GET /api/users/verify-email` — Verify email via token
  - `POST /api/users/logout` — Clear session cookie
- Integrated:
  - JWT-based authentication using HTTP-only cookies
  - Token generation and validation for email verification and password reset
  - Session handling and middleware support
- Added secure token expiry and usage checks

#### Frontend Implementation

- Created pages for:
  - Signup
  - Login
  - Forgot Password
  - Reset Password
  - Email Verification feedback
- Form validation and state handling using `useState` and controlled components
- Added user feedback with `react-hot-toast`
- Redirects based on login/verification state
- Error handling for invalid tokens, wrong credentials, and expired links

#### Session & Security

- Secure JWT handling via cookies
- Token expiry and single-use protection
- Logout clears session and redirects appropriately
---

jishnu-soundspire commented:
Can we change all occurrences of **isVerified** to **is_verified** as that is what we have in the DB schema and also what I have in the local DB setup? @skjaiswal88 and @saumyatalwani please confirm if you guys have isVerified or is_verified; it's okay if you don't have it setup.

skjaiswal88 commented:
> Can we change all occurrences of **isVerified** to **is_verified** as that is what we have in the DB schema and also what I have in the local DB setup? @skjaiswal88 and @saumyatalwani please confirm if you guys have isVerified or is_verified; it's okay if you don't have it setup.

i have is_verified in my table

jishnu-soundspire commented:
As everyone has is_verified, lets keep that and not isVerified @killlercram

killlercram commented on src/app/pageBackup.tsx:
```diff
@@ -0,0 +1,241 @@
+  useEffect(() => {
+    const allFilled = Object.values(user).every((val) => val.trim().length > 0);
+    setButtonDisabled(!allFilled);
+    // if (
+    //   user.username.length > 0 &&
+    //   user.email.length > 0 &&
+    //   user.password_hash.length > 0 &&
+    //   user.full_name.length > 0 &&
+    //   user.gender.length > 0 &&
+    //   user.mobile_number.length > 0 &&
+    //   user.date_of_birth.length > 0 &&
+    //   user.city.length > 0
+    // ) {
+    //   setButtonDisabled(false);
+    // } else {
+    //   setButtonDisabled(true);
+    // }
```
These were for the frontend form validation, which are not in use now as i used better validation then these.So i will simply remove these.

killlercram commented on src/app/api/auth/session/route.ts:
```diff
@@ -1,22 +1,64 @@
+    await connectionTestingAndHelper();
+
+
+    // const token = (await cookies()).get("token")?.value;
+    // if(!token) return NextResponse.json({user:null});
+
+    // const decode = jwt.verify(token, process.env.JWT_SECRET!) as DecodedToken;
+    // const user = await User.findOne({where: {user_id: decode.id}});
+
+    // if(!user) return NextResponse.json({user:null});
+
+    // return NextResponse.json({
+    //   user: {
+    //     id: user.user_id,
+    //     name: user.full_name,
+    //     email: user.email,
+    //     provider: "local",
+    //   },
+    // });
```
These lines are also of no use it can also be removed.

killlercram commented on src/app/(protected)/explore/page.tsx:
```diff
@@ -1,11 +1,14 @@
-import ArtistCard from '@/components/ArtistCard';
-import ReviewCard from '@/components/ReviewCard';
-import GenreCard from '@/components/GenreCard';
+// import Carousel from '@/components/Carousel';
+// import ArtistCard from '@/components/ArtistCard';
+// import ReviewCard from '@/components/ReviewCard';
+// import GenreCard from '@/components/GenreCard';
```
I have commented these out as these are declared but never been used in the explore pages so i have commented these out once the pages being made they can be used by uncommenting these or even we can remove these until we create these.So its not an issue.

jishnu-soundspire replied:
ok

killlercram commented:
> Can we change all occurrences of **isVerified** to **is_verified** as that is what we have in the DB schema and also what I have in the local DB setup? @skjaiswal88 and @saumyatalwani please confirm if you guys have isVerified or is_verified; it's okay if you don't have it setup.

"I think the data base which was provided has is_verified only."

jishnu-soundspire commented:
> ![WhatsApp Image 2025-06-25 at 7 44 44 PM](https://private-user-images.githubusercontent.com/111636617/458914030-f1dd5386-0263-4cd8-9663-9af3888769e3.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA5NTI5OTgsIm5iZiI6MTc1MDk1MjY5OCwicGF0aCI6Ii8xMTE2MzY2MTcvNDU4OTE0MDMwLWYxZGQ1Mzg2LTAyNjMtNGNkOC05NjYzLTlhZjM4ODg3NjllMy5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MjZUMTU0NDU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWM2MmY5YjM3NjU2Y2VlYjk0MDIxYzI5ZTYyODJlNmExNmNkMTkwNmRkOTE0OTkzODhjODBkMTRmMGI4N2FmOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.0WlZJFCdnO0VKmK_EYIRtn1Wf2CqaHT0YSfcxMHIM2s)
> 
> > Can we change all occurrences of **isVerified** to **is_verified** as that is what we have in the DB schema and also what I have in the local DB setup? @skjaiswal88 and @saumyatalwani please confirm if you guys have isVerified or is_verified; it's okay if you don't have it setup.
> 
> I think the data base which was provided has "is_verified" only, so i think you all must change??.

Haan, True, it should be is_verified only, but in your PR it looks like it is isVerified and not is_verified.
Ref: https://github.com/Tech-SoundSpire/soundspire-frontend/pull/15/files#diff-0a3f878b49e77587554902a63f4b05fbf91ef3f87a627edc34d33e69f0471764R57

killlercram commented:
Removed all the Comments as asked change isVerified to is_verified.

killlercram commented:
Removed all the Comments as asked change isVerified to is_verified.

killlercram commented:
> Got this error when trying to login ![image](https://private-user-images.githubusercontent.com/203329727/459997636-538e3452-1089-4e1e-9f36-14aaa085145c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTExMDMxNjgsIm5iZiI6MTc1MTEwMjg2OCwicGF0aCI6Ii8yMDMzMjk3MjcvNDU5OTk3NjM2LTUzOGUzNDUyLTEwODktNGUxZS05ZjM2LTE0YWFhMDg1MTQ1Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyOFQwOTI3NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MzI1Zjc2MGM1NzgwMWU5Zjk0NjE4NzY3OTk0MmRjNzVmZTE2YzM3ZGMzZjE4YzRiNzdhMTI0ZTE0MDBkZThiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.SKkWXMNPhVfeFLXAd3A2aVgVO6AZs2vhcQk60HqhepY)
> 
> Got another one when I clicked on forgot password:
> 
> ```
> Forgot password Error: secretOrPrivateKey must have a value
>  POST /api/users/forgot-password 500 in 2703ms
> Forgot password Error: secretOrPrivateKey must have a value
>  POST /api/users/forgot-password 500 in 226ms
> ```
> 
> For this one, I think I am missing something. Please educate me on what I am missing.
> 
> Got the same when signing up for a new user:
> 
> ```
>  POST /api/users/signup 500 in 3345ms
> All models were synchronized successfully!
> {
>  full_name: 'Jishnu kS',
>  username: 'jks',
>  email: 'jisheras31@gmail.com',
>  password_hash: 'jishnu',
>  gender: 'male',
>  mobile_number: '1234567890',
>  date_of_birth: '2001-08-31',
>  city: 'Pune'
> }
> 🔍 Checking for existing user...
> 🔐 Hashing password...
> 🧑‍💻 Creating new user...
> POST /api/users/signup 500 in 258ms
> ```
> 
> The rest looks good more or less.

Looking into it

killlercram commented:
I have tested and i think it broke because i changed some of the things in the profile page now i reverted back the changes and went to the previous code and everything is working fine.It is properly Signing up, logging in and Forget and reset is also working fine.
Attaching the testings and pushing the latest code soon.
![Screenshot from 2025-06-28 17-20-42](https://github.com/user-attachments/assets/dd23c051-0a69-4666-ad0f-8203ad8cfacf)
![Screenshot from 2025-06-28 17-23-04](https://github.com/user-attachments/assets/6408c111-97a4-4e43-926a-a38a6b50e0b8)
![Screenshot from 2025-06-28 17-23-48](https://github.com/user-attachments/assets/b461ded2-7e42-41b6-a7f5-f987e146889b)
![Screenshot from 2025-06-28 17-24-15](https://github.com/user-attachments/assets/59a6dbe1-cefb-4034-86d6-ea04b2150c9e)
![Screenshot from 2025-06-28 17-24-59](https://github.com/user-attachments/assets/55f59a12-5a20-4a4b-9f5c-1d2915fab651)
![Screenshot from 2025-06-28 17-25-26](https://github.com/user-attachments/assets/7926cf71-ef6f-4da8-8960-b4c8b7168f52)

killlercram commented:
These are the Environment variables used:

#SIGN_UP/IN
DOMAIN=https://localhost:3000
DB_NAME=""
DB_USER=""
DB_PASSWORD=""
DB_HOST="localhost"
DB_PORT=5432
DB_DIALECT=postgres
JWT_SECRET=


#MailTrap for Sending Mail
MAILTRAP_HOST=
MAILTRAP_PORT=
MAILTRAP_USER=
MAILTRAP_PASSWORD=

jishnu-soundspire reviewed (approved):
Looks good, GG
